core/src/test/java/com/linecorp/armeria/common/http/PathParamExtractorTest.java

```diff
@@ -0,0 +1,87 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common.http;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.Map;
+
+import org.junit.Test;
+
+public class PathParamExtractorTest {
+    @Test
+    public void givenMatchingPathParam_whenExtract_thenReturns() throws Exception {
+        PathParamExtractor ppe = PathParamExtractor.of("/service/{value}");
+
+        Map<String, String> values = ppe.extract("/service/hello");
+
+        assertThat(values).hasSize(1);
+        assertThat(values.get("value")).isEqualTo("hello");
+    }
+
+    @Test
+    public void givenNoMatchingPathParam_whenExtract_thenReturnsEmptyMap() throws Exception {
+        PathParamExtractor ppe = PathParamExtractor.of("/service/{value}");
+
+        Map<String, String> values = ppe.extract("/service2/hello");
+
+        assertThat(values).isEmpty();
+    }
+
+    @Test
+    public void testMultipleMatch() throws Exception {
+        PathParamExtractor ppe = PathParamExtractor.of("service/{value}/test/:value2/something");
+
+        Map<String, String> values = ppe.extract("service/hello/test/world/something?q=1");
+
+        assertThat(values).hasSize(2);
+        assertThat(values.get("value")).isEqualTo("hello");
+        assertThat(values.get("value2")).isEqualTo("world");
+    }
+
+    @Test
+    public void testVariables() throws Exception {
+        PathParamExtractor ppe = PathParamExtractor.of("/service/{value}/test/:value2/something/{value3}");
+
+        assertThat(ppe.variables()).containsExactlyInAnyOrder("value", "value2", "value3");
+    }
+
+    @Test
+    public void testSkeleton() throws Exception {
+        PathParamExtractor ppe = PathParamExtractor.of("/service/{value}/test/:value2/something/{value3}");
+
+        assertThat(ppe.skeleton()).isEqualTo("/service/{}/test/{}/something/{}");
+    }
+
+    @Test
+    public void testInvalidPattern() throws Exception {
+        assertThatThrownBy(() -> PathParamExtractor.of("/service/{value}/test/{value2"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("/service/{value}/test/{value2: Invalid PathParamExtractor");
+
+        assertThatThrownBy(() -> PathParamExtractor.of("/service/:value/test/value2}"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("/service/:value/test/value2}: Invalid PathParamExtractor");
+    }
+
+    @Test
+    public void testEmptyPattern() throws Exception {
+        assertThatThrownBy(() -> PathParamExtractor.of("/service/value/test/value2"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("No placeholder exists in given path: /service/value/test/value2");
```

Comment:
I'm curious if a path pattern without variable placeholders should be rejected. Perhaps OK to return an empty map on `extract()`?